File: algorithms/javaalgo/src/test/java/programmers/level2/fatigue/FatigueTest.java
package programmers.level2.fatigue;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

public class FatigueTest {

    static boolean[] visited;
    static int answer;

    @Test
    void test() {
        int k = 80;
        int[][] dungeons = {
            {80, 20},
            {50, 40},
            {30, 10}
        };

        assertThat(solution(k, dungeons)).isEqualTo(3);
    }

    public int solution(int k, int[][] dungeons) {
        answer = 0;
        visited = new boolean[dungeons.length];

        dfs(0, k, dungeons);

        return answer;
    }

    public void dfs(int L, int k, int[][] dungeons) {
        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && dungeons[i][0] <= k) {
                visited[i] = true;
                dfs(L + 1, k - dungeons[i][1], dungeons);
                visited[i] = false;
            }
            answer = Math.max(answer, L);
        }
    }
}

File: algorithms/javascriptalgo/programmers/level2/fatigue.test.js
test("test", () => {
    const k = 80;
    const dungeons = [[80, 20], [50, 40], [30, 10]];

    expect(solution(k, dungeons)).toBe(3);
})

function solution(k, dungeons) {
    let answer = 0;
    let visited = new Array(dungeons.length).fill(false);

    function dfs(L, k, dungeons) {
        for (let i = 0; i < dungeons.length; i++) {
            if (!visited[i] && dungeons[i][0] <= k) {
                visited[i] = true;
                dfs(L + 1, k - dungeons[i][1], dungeons);
                visited[i] = false;
            }
            answer = Math.max(answer, L);
        }
    }

    dfs(0, k, dungeons);

    return answer;
}
